src/utils/SSR.ts

```diff
@@ -0,0 +1,96 @@
+import { NormalizedCacheObject } from 'apollo-cache-inmemory';
+import { ApolloClient } from 'apollo-client';
+import { IncomingMessage } from 'http';
+import get from 'lodash.get';
+import NextCookies from 'next-cookies';
+import { LAYOUT_QUERY } from '../gql/common/layoutQuery';
+import { ApolloQueryOptions } from '../types/ApolloQueryOptions';
+import { Cookies } from '../types/Cookies';
+import { GetServerSidePropsContext } from '../types/GetServerSidePropsContext';
+import { PublicHeaders } from '../types/PublicHeaders';
+import { UniversalSSRPageProps } from '../types/UniversalSSRPageProps';
+import { UserSemiPersistentSession } from '../types/UserSemiPersistentSession';
+import { prepareGraphCMSLocaleHeader } from './graphcms';
+import { createApolloClient } from './graphql';
+import { DEFAULT_LOCALE, resolveFallbackLanguage } from './i18n';
+import { fetchTranslations, I18nextResources } from './i18nextLocize';
+import UniversalCookiesManager from './UniversalCookiesManager';
+
+/**
+ * getCommonServerSideProps returns only part of the props expected in UniversalSSRPageProps
+ * To avoid TS issue, we omit those that we don't return, and add those necessary to the getServerSideProps function
+ */
+export type getCommonServerSidePropsResults = Omit<UniversalSSRPageProps, 'apolloState' | 'customer'> & {
+  apolloClient: ApolloClient<NormalizedCacheObject>;
+  layoutQueryOptions: ApolloQueryOptions;
+  headers: PublicHeaders;
+}
+
+/**
+ * Only executed on the server side, for every request.
+ * Computes some dynamic props that should be available for all SSR pages that use getServerSideProps
+ *
+ * Because the exact GQL query will depend on the consumer, this helper doesn't run any query by itself, but rather return all necessary props to allow the consumer to perform its own queries
+ * This improves performances, by only running one GQL query instead of many (consumer's choice)
+ *
+ * Meant to avoid code duplication
+ *
+ * @param context
+ *
+ * @see https://nextjs.org/docs/basic-features/data-fetching#getserversideprops-server-side-rendering
+ */
+export const getCommonServerSideProps = async (context: GetServerSidePropsContext): Promise<getCommonServerSidePropsResults> => {
```

Comment:
Loved it!

Reply:
It's a first draft but I believe it provides enough flexibility for any real use-case. Wasn't obvious how to do that considering perfs + code reusability, don't hesitate to let me know if stuff is missing